plugins/samples/enable_recaptcha/plugin.rs

```diff
@@ -0,0 +1,225 @@
+    // and modifies chunks of Html. Behavior for Html modifications are defined
+    // via content_handlers. The rewriter only writes completed sections of Html to
+    // the output sink(e.g., when the rewriter parses "<di", nothing will be
+    // sent to the output sink. If the next chunk was "v> <h1>foo</h", the
+    // rewriter would then write "<div><h1>foo" to the output sink) unless
+    // HtmlRewriter::end() is called. HtmlRewriter::end() results in all
+    // uncompleted Html being buffered in rewriter to be sent to output sink as
+    // if it were plain text.
+    rewriter: Option<HtmlRewriter<'a, MyOutputSink>>,
+    // True when plugin has added script to <head>.
+    completed_script_injection: Rc<RefCell<bool>>,
+    // reCaptcha site key
+    recaptcha_key: Rc<RefCell<String>>,
+}
+
+impl<'a> MyHttpContext<'a> {
+    pub fn new(key: Rc<RefCell<String>>) -> MyHttpContext<'a> {
+        let output = Rc::new(RefCell::new(Vec::new()));
+        let completed_script_injection = Rc::new(RefCell::new(false));
+        let recaptcha_key = key;
+        MyHttpContext {
+            output: output.clone(),
+            completed_script_injection: completed_script_injection.clone(),
+            recaptcha_key: recaptcha_key.clone(),
+            rewriter: Some(HtmlRewriter::new(
+                Settings {
+                    element_content_handlers: vec![element!("head", move |el| {
+                        el.prepend(
+                            format!("\n<script src=\"https://www.google.com/recaptcha/enterprise.js?render={}\"></script>\n", *recaptcha_key.borrow_mut()).as_str(),
+                            ContentType::Html,
+                            );
+                        *completed_script_injection.borrow_mut() = true;
+                        Ok(())
+                    })],
+                    ..Settings::new()
+                },
+                MyOutputSink {
+                    output_sink: output,
+                },
+            )),
+        }
+    }
+
+    fn parse_chunk(&mut self, body_bytes: Bytes) -> Result<(), Box<dyn Error>> {
+        // Parse/rewrite current chunk
+        self.rewriter
+            .as_mut()
+            .ok_or("Expected valid rewriter. Got None")?
+            .write(&body_bytes)?;
+        return Ok(());
+    }
+
+    fn end_rewriter(&mut self) -> Result<(), Box<dyn Error>> {
+        // Stop the rewriter after completing desired domain rewrite to
+        // dump any unparsable inputs to output.
+        self.rewriter
+            .take()
+            .expect("Expected valid rewriter. Got None")
+            .end()?;
+        return Ok(());
+    }
+}
+
+impl<'a> Context for MyHttpContext<'a> {}
+
+impl<'a> HttpContext for MyHttpContext<'a> {
+    // Use header 'stop-compression' to toggle body compression.
+    fn on_http_request_headers(&mut self, _: usize, _: bool) -> Action {
+        let stop_compression = self.get_http_request_header("Stop-compresssion");
```

Comment:
Strongly prefer this is in the plugin configuration rather than forcing users to modify web requests from their client to include this header. Fine for now I guess.

Reply:
Ack

Reply:
I agree with @jordan83. This doesn't need to be a header-driven conditional. In the Service Extension resources configuration, a user may choose to invoke this request headers handler or not, so I propose we just remove the use of "Stop-compresssion" here, and replace it with a comment saying "invoke this handler to disable compression".

Reply:
Decided to ditch the disable-compression feature, please remove.